utilities/src/fixed_128.rs

```diff
@@ -0,0 +1,509 @@
+use codec::{Decode, Encode};
+use primitives::U256;
+use rstd::convert::{Into, TryFrom, TryInto};
+use sp_runtime::{
+	traits::{Bounded, Saturating, UniqueSaturatedInto},
+	PerThing,
+};
+
+#[cfg(feature = "std")]
+use serde::{de, Deserialize, Deserializer, Serialize, Serializer};
+use sp_runtime::traits::SaturatedConversion;
+
+/// A signed fixed-point number. Can hold any value in the range [-170_141_183_460_469_231_731, 170_141_183_460_469_231_731]
+/// with fixed-point accuracy of 10 ** 18.
+#[derive(Encode, Decode, Default, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Fixed128(i128);
+
+const DIV: i128 = 1_000_000_000_000_000_000;
+
+impl Fixed128 {
+	/// Create self from a natural number.
+	///
+	/// Note that this might be lossy.
+	pub fn from_natural(int: i128) -> Self {
+		Self(int.saturating_mul(DIV))
+	}
+
+	/// Accuracy of `Fixed128`.
+	pub const fn accuracy() -> i128 {
+		DIV
+	}
+
+	/// Raw constructor. Equal to `parts / DIV`.
+	pub fn from_parts(parts: i128) -> Self {
+		Self(parts)
+	}
+
+	/// Creates self from a rational number. Equal to `n/d`.
+	///
+	/// Note that this might be lossy.
+	pub fn from_rational<N: UniqueSaturatedInto<i128>>(n: N, d: N) -> Self {
+		// TODO: Should have a better way to enforce this requirement
+		let n = n.unique_saturated_into();
+		let d = d.unique_saturated_into();
+		Self(
+			(n.saturating_mul(DIV.into()) / d.max(1))
```

Comment:
If `d` is negative, `d.max(1)` will always get `1`. And we need a test case to cover it, like:
```rust
let a = Fixed128::from_rational(2, -1);
assert_eq!(a, Fixed128::from_natural(-2));
```

Reply:
@shaopengw making d: u128 until we find a better way